crates/typst-library/src/layout/grid/resolve.rs

```diff
@@ -1692,24 +1864,76 @@ fn resolve_cell_position(
             }
             if let Smart::Custom(cell_y) = cell_y {
                 // Cell has chosen its exact position.
+                //
+                // Ensure it doesn't conflict with an existing header or
+                // footer (but only if it isn't already in one, otherwise there
+                // will already be a separate check).
+                if !is_row_group {
+                    check_for_conflicting_cell_row(header, footer, cell_y, rowspan)?;
+                }
+
                 cell_index(cell_x, cell_y)
             } else {
                 // Cell has only chosen its column.
                 // Let's find the first row which has that column available.
-                let mut resolved_y = 0;
-                while let Some(Some(_)) =
-                    resolved_cells.get(cell_index(cell_x, resolved_y)?)
-                {
-                    // Try each row until either we reach an absent position
-                    // (`Some(None)`) or an out of bounds position (`None`),
-                    // in which case we'd create a new row to place this cell in.
-                    resolved_y += 1;
+                // If in a header or footer, start searching by the first empty
+                // row / the header or footer's first row (specified through
+                // 'first_available_row'). Otherwise, start searching at the
+                // first row.
+                let mut resolved_y = first_available_row;
+
+                // There may be row groups, so we have to not only skip
+                // rows where the requested column is occupied to find the
+                // first suitable row, but also skip rows belonging to
+                // headers or footers.
+                loop {
+                    if let Some(Some(_)) =
+                        resolved_cells.get(cell_index(cell_x, resolved_y)?)
```

Comment:
Should take the deduplication opportunity to not call `cell_index(...)?` each time and instead start at a suitable index and keep trying to increase it by 1 (top loop) / `columns` (bottom - this - loop)